config/v1/types_dns.go

```diff
@@ -26,6 +26,25 @@ type DNSSpec struct {
 	// For example, given the base domain `openshift.example.com`, an API server
 	// DNS record may be created for `cluster-api.openshift.example.com`.
 	BaseDomain string `json:"baseDomain"`
+	// publicZone is the location where all the DNS records that are publicly accessible to
+	// the internet exist.
+	// If this field is nil, no public records should be created.
+	// +optional
+	PublicZone *DNSZone `json:"publicZone,omitempty"`
+	// privateZone is the location where all the DNS records that are only available internally
+	// to the cluster exist.
+	PrivateZone DNSZone `json:"privateZone"`
+}
+
+// DNSZone is used to define a DNS hosted zone.
+// A zone can be identified by an ID or tags.
+type DNSZone struct {
+	// id is the identifier that can be used to find the DNS hosted zone.
+	// +optional
+	ID *string `json:"id"`
+	// tags can be used to query the DNS hosted zone.
```

Comment:
Tags is dangerous, that's not going to be identical across platforms.  What is this actually capturing?

Reply:
If this isn't truly platform independent you need to make this specific to the platform, or make the details opaque (i.e. have a string defined as opaque that no one except the person who wrote it is allowed to read).

Reply:
The installer currently cannot use ID to define the private r53 zone as it is created by the private zone is created by terraform.. so ID is not known when `DNS` object is created by installer. We certainly know the tags that will exist on that private zone.

Therefore DNSZone has `ID` and `Tags`, where `one of them` is going to be set to find the zone.

And I thought finding a resource using `ID` or `Tags` is something all platforms support?
AWS ID: [id](https://docs.aws.amazon.com/cli/latest/reference/route53/get-hosted-zone.html#options) Tags: [tags](https://docs.aws.amazon.com/cli/latest/reference/resourcegroupstaggingapi/get-resources.html#options)
Azure ID: [name](https://docs.microsoft.com/en-us/cli/azure/network/dns/zone?view=azure-cli-latest#az-network-dns-zone-show)
GCP ID: [name/id](https://cloud.google.com/dns/docs/reference/v1/managedZones/get)

Reply:
If the docs for a human setting this would have to explain it, then it should probably be multiple fields which cannot all be set together (union type) like `AWSID, AWSTags, AzureName, GCPName` etc.

We intend for end users to set this?  What happens when a human changes this?  Which operator(s) react and make it valid?  If we think that this is set once at install time, and we don't envision a human changing (now, a human could override it maybe later) then status may be more appropriate.

Reply:
> ... like `AWSID`, `AWSTags`, `AzureName`, `GCPName` etc.

If "ID strings" are a cross-platform concept (and they are, right?), then I don't see a need for platform-specific versions of ID property.  Same for tags, although from @abhinavdahiya's comment above it sounds like tags may not be cross-platform, with some platforms supporting tags and some supporting creator supplied names.

> We intend for end users to set this?

I don't see a problem with users setting tags, but I don't think users will need that functionality either.  And while it's theoretically possible to suck Route 53 zone creation up into the asset graph so we could fill in the ID here, that would break the current, useful separation between "create all the assets locally" and "push them up to the target cloud" that BYOH and similar rely on (I think?).

> What happens when a human changes this? Which operator(s) react and make it valid?

I don't think this is different from the ID.  In both cases, the ingress operator is pulling fresh values from the config and using them to look up the zones under management.

Reply:
> > What happens when a human changes this? Which operator(s) react and make it valid?
> 
> I don't think this is different from the ID. In both cases, the ingress operator is pulling fresh values from the config and using them to look up the zones under management.

The operator can use a watch on the config object to react when `ID` or `Tags` changes in the config object, but it would be trickier to detect changes when an administrator modifies tags on the hosted zones—AFAICT, that would involve polling.  If the operator is supposed to react to this sort of configuration change, then using `ID` will make the implementation more reactive and less burdensome in terms of AWS API requests.

Reply:
So treating `ID` and `Tags` as the basis of a platform-agnostic union type seems to cover all of the providers mentioned so far. Is the fear that we'll discover another case later that doesn't fit the model, forcing us to walk it back into platform-specific types anyway? What other concerns are there with ID+Tags?

Reply:
I think we should move with `ID/Tags` for now. since it can cover most cloud platforms with programmable DNS.

Reply:
That's fine.

Reply:
Please add slightly more godoc here capturing this discussion and how it should be used.